src/stan/io/stan_csv_reader.hpp

```diff
@@ -348,36 +351,55 @@ class stan_csv_reader {
   /**
    * Parses the file.
    *
+   * Throws exception if contents can't be parsed into header + data rows
+   * or if sample size doesn't match metadata config.
+   *
+   * Emits warning message if can't parse sampler adaptation.
+   *
    * @param[in] in input stream to parse
    * @param[out] out output stream to send messages
    */
   static stan_csv parse(std::istream& in, std::ostream* out) {
     stan_csv data;
+    std::string line;
 
-    if (!read_metadata(in, data.metadata, out)) {
-      if (out)
-        *out << "Warning: non-fatal error reading metadata" << std::endl;
+    read_metadata(in, data.metadata);
+    if (!read_header(in, data.header)) {
+      throw std::invalid_argument("Error with header of input file in parse");
     }
 
-    if (!read_header(in, data.header, out)) {
-      if (out)
-        *out << "Error: error reading header" << std::endl;
-      throw std::invalid_argument("Error with header of input file in parse");
+    // skip warmup draws, if any
+    if (data.metadata.algorithm != "fixed_param" && data.metadata.num_warmup > 0
+        && data.metadata.save_warmup) {
+      while (in.peek() != '#') {
+        std::getline(in, line);
+      }
     }
 
-    if (!read_adaptation(in, data.adaptation, out)) {
-      if (out)
-        *out << "Warning: non-fatal error reading adaptation data" << std::endl;
+    if (data.metadata.algorithm != "fixed_param") {
+      read_adaptation(in, data.adaptation);
     }
 
     data.timing.warmup = 0;
     data.timing.sampling = 0;
 
+    if (data.metadata.method == "variational") {
+      std::getline(in, line);  // discard variational estimate
+    }
+
     if (!read_samples(in, data.samples, data.timing, out)) {
       if (out)
         *out << "Warning: non-fatal error reading samples" << std::endl;
     }
-
+    if (data.metadata.method == "sample") {
+      int expected_samples = data.metadata.num_samples / data.metadata.thin;
+      if (expected_samples != data.samples.rows()) {
+        std::stringstream msg;
+        msg << "Error reading samples, expecting " << expected_samples
+            << " samples, found " << data.samples.rows();
+        throw std::invalid_argument(msg.str());
```

Comment:
I'm not sure of the value add of checking this -- in particularly, it would prevent us from summarizing a csv file that didn't come from Stan, which is otherwise fine (and useful, I would argue).

Because we're now setting the `badbit` exception flag (https://github.com/stan-dev/cmdstan/pull/1249) we don't need to worry as much about malformed CSVs like this going undetected

Reply:
removed.